demos/generomak_solps_from_raw_files.py

```diff
@@ -0,0 +1,234 @@
+# Load the simulation.
+###############################################################################
+demos_directory = os.path.dirname(__file__)
+simulation_directory = os.path.join(demos_directory, 'generomak_raw_files')
+try:
+    print('Loading simulation...')
+    sim = load_solps_from_raw_output(simulation_directory)
+except RuntimeError:
+    print('ERROR: simulation files not found. Please extract the raw '
+          'files from the zip archive before running this demo:')
+    print('> unzip {0}/generomak_raw_files.zip -d {0}/generomak_raw_files'.format(demos_directory))
+    sys.exit(1)
+
+###############################################################################
+# Create a Cherab plasma from the simulation
+###############################################################################
+mesh = sim.mesh
+plasma = sim.create_plasma()
+
+me = mesh.mesh_extent
+xl, xu = (me['minr'], me['maxr'])
+zl, zu = (me['minz'], me['maxz'])
+nx = 500
+ny = 500
+
+# Extract information about the species in the plasma.
+h0 = plasma.composition.get(hydrogen, 0)
+h1 = plasma.composition.get(hydrogen, 1)
+c0 = plasma.composition.get(carbon, 0)
+c1 = plasma.composition.get(carbon, 1)
+c2 = plasma.composition.get(carbon, 2)
+c3 = plasma.composition.get(carbon, 3)
+c4 = plasma.composition.get(carbon, 4)
+c5 = plasma.composition.get(carbon, 5)
+c6 = plasma.composition.get(carbon, 6)
+
+###############################################################################
+# Sample some poloidal profiles from the simulation.
+###############################################################################
+print('Sampling profiles...')
+# The distributions are 3D, so we perform a sample in 3D space with only a
+# single y value to get a poloidal profile.
+xsamp, _, zsamp, ne_samples = sample3d(plasma.electron_distribution.density,
+                                       (xl, xu, nx), (0, 0, 1), (zl, zu, ny))
+ne_samples = ne_samples.squeeze()
+te_samples = sample3d_grid(plasma.electron_distribution.effective_temperature,
+                           xsamp, [0], zsamp).squeeze()
+h0_samples = sample3d_grid(h0.distribution.density, xsamp, [0], zsamp).squeeze()
+h1_samples = sample3d_grid(h1.distribution.density, xsamp, [0], zsamp).squeeze()
+c0_samples = sample3d_grid(c0.distribution.density, xsamp, [0], zsamp).squeeze()
+c1_samples = sample3d_grid(c1.distribution.density, xsamp, [0], zsamp).squeeze()
+c2_samples = sample3d_grid(c2.distribution.density, xsamp, [0], zsamp).squeeze()
+c3_samples = sample3d_grid(c3.distribution.density, xsamp, [0], zsamp).squeeze()
+c4_samples = sample3d_grid(c4.distribution.density, xsamp, [0], zsamp).squeeze()
+c5_samples = sample3d_grid(c5.distribution.density, xsamp, [0], zsamp).squeeze()
+c6_samples = sample3d_grid(c6.distribution.density, xsamp, [0], zsamp).squeeze()
+# Magnitude of main ion velocity vector.
+h1_velocity = samplevector3d_grid(h1.distribution.bulk_velocity, xsamp, [0], zsamp).squeeze()
```

Comment:
With the new interpolators, these distributions can be sampled easier.
```suggestion
xsamp, zsamp, ne_samples = sample2d(sim.electron_density_f2d, (xl, xu, nx), (zl, zu, ny))
ne_samples = ne_samples.squeeze()
te_samples = sample2d_grid(sim.electron_temperature_f2d, xsamp, zsamp).squeeze()
h0_samples = sample2d_grid(sim.species_density_f2d[('hydrogen', 0)], xsamp, zsamp).squeeze()
h1_samples = sample2d_grid(sim.species_density_f2d[('hydrogen', 1)], xsamp, zsamp).squeeze()
c0_samples = sample2d_grid(sim.species_density_f2d[('carbon', 0)], xsamp, zsamp).squeeze()
c1_samples = sample2d_grid(sim.species_density_f2d[('carbon', 1)], xsamp, zsamp).squeeze()
c2_samples = sample2d_grid(sim.species_density_f2d[('carbon', 2)], xsamp, zsamp).squeeze()
c3_samples = sample2d_grid(sim.species_density_f2d[('carbon', 3)], xsamp, zsamp).squeeze()
c4_samples = sample2d_grid(sim.species_density_f2d[('carbon', 4)], xsamp, zsamp).squeeze()
c5_samples = sample2d_grid(sim.species_density_f2d[('carbon', 5)], xsamp, zsamp).squeeze()
c6_samples = sample2d_grid(sim.species_density_f2d[('carbon', 6)], xsamp, zsamp).squeeze()
# Magnitude of main ion velocity vector.
h1_velocity = samplevector2d_grid(sim.velocities_cylindrical_f2d[('hydrogen', 1)], xsamp, zsamp).squeeze()
```
However, if the goal of this demo is to show how the plasma object is built from the simulation, then it's better to leave it as it is.

Reply:
This really depends on the the aim of the demo. If this is to be aiming at the solps package, then I would use what @vsnever suggests. The option with plasma distribution can be used in a demo more aiming at the plasma class.

Reply:
Yes, this is open to interpretation. Ultimately users are likely to be using Cherab for its plasma object, and not just an alternative to `b2plot`. But verifying the SOLPS data has been loaded correctly is also important.

Perhaps we should have two similar demos, one which samples from the simulation object and one which samples from the plasma object. What do you think?

Reply:
I would show different ways of sampling in the same demo and add comments explaining that there are a few ways to do this.

Reply:
OK, sounds good. I'll put in something like `np.all(sim_samples == plasma_samples)` to demonstrate, rather than duplicating all the plots. And I won't use both methods for all the carbon ionisation stages, but can show the two methods for ne, Te, nC0 for example.